easycv/datasets/detection/data_sources/base.py

```diff
@@ -0,0 +1,171 @@
+
+
+def load_image(img_path):
+    result = {}
+    try_cnt = 0
+    img = None
+    while try_cnt < MAX_READ_IMAGE_TRY_TIMES:
+        try:
+            with io.open(img_path, 'rb') as infile:
+                # cv2.imdecode may corrupt when the img is broken
+                image = Image.open(infile)
+                img = cv2.cvtColor(
+                    np.asarray(image, dtype=np.uint8), cv2.COLOR_RGB2BGR)
+                assert img is not None, 'Image load error, try %s : %s' % (
+                    try_cnt, img_path)
+                break
+        except:
+            time.sleep(2)
+        try_cnt += 1
+
+    if img is None:
+        raise ValueError('Read Image Times Out: ' + img_path)
+
+    result['img'] = img.astype(np.float32)
+    result['img_shape'] = img.shape  # h, w, c
+    result['ori_img_shape'] = img.shape
+
+    return result
+
+
+def build_sample(source_item, classes, parse_fn, load_img):
+    """Build sample info from source item.
+    Args:
+        source_item: item of source iterator
+        classes: classes list
+        parse_fn: parse function to parse source_item, only accepts two params: source_item and classes
+        load_img: load image or not, if true, cache all images in memory at init
+    """
+    result_dict = parse_fn(source_item, classes)
+
+    if load_img:
+        result_dict.update(load_image(result_dict['filename']))
+
+    return result_dict
+
+
+class DetSourceBase(object):
+
+    def __init__(self,
+                 classes=[],
+                 cache_at_init=False,
+                 cache_on_the_fly=False,
+                 parse_fn=None,
+                 num_processes=int(cpu_count() / 2),
+                 **kwargs):
+        """
+        Args:
+            classes: classes list
+            cache_at_init: if set True, will cache in memory in __init__ for faster training
+            cache_on_the_fly: if set True, will cache in memroy during training
+            parse_fn: parse function to parse source iterator
+            num_processes: number of processes to parse samples
+        """
+        self.CLASSES = classes
+        self.rank, self.world_size = get_dist_info()
+        self.cache_at_init = cache_at_init
+        self.cache_on_the_fly = cache_on_the_fly
+        self.num_processes = num_processes
+
+        if self.cache_at_init and self.cache_on_the_fly:
+            raise ValueError(
+                'Only one of `cache_on_the_fly` and `cache_at_init` can be True!'
+            )
+        source_iter = self.get_source_iterator()
+
+        process_fn = functools.partial(
+            build_sample,
+            parse_fn=parse_fn,
+            classes=self.CLASSES,
+            load_img=cache_at_init == True,
+        )
+        self.samples_list = self.build_samples(
+            source_iter, process_fn=process_fn)
+
+    @abstractmethod
+    def get_source_iterator():
+        """data list iterator, for multi-process read
```

Comment:
more detailed docstr is missing, iterator of which kind format of data

Reply:
done!